source/Random.h

```diff
@@ -39,8 +39,9 @@ class Random {
 	static uint32_t Polya(uint32_t k, double p = .5);
 	// Get a number from a binomial distribution (i.e. integer bell curve).
 	static uint32_t Binomial(uint32_t t, double p = .5);
-	// Get a normally distributed number (mean = 0, sigma= 1).
-	static double Normal();
+	// Get a number from a normal distribution with default or specified mean and stddev using the Box-Muller transform.
+	// Also caches the unused variable without transforming it.
+	static double BoxMullerSample(double mean = 0, double sigma = 1);
```

Comment:
I benchmarked the implementation of `Random::Normal()` on the main branch 4 times:
```
benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Random::Normal                                 100           316     2.5596 ms 
                                        80.1059 ns     79.498 ns    80.7632 ns 
                                        3.22394 ns    2.82655 ns    4.26721 ns 
Random::Normal                                 100           334     2.7054 ms 
                                        80.3474 ns    79.7631 ns    81.1615 ns 
                                         3.4813 ns     2.6076 ns    5.81454 ns 
Random::Normal                                 100           305      2.562 ms 
                                        80.9188 ns    79.9836 ns    82.7276 ns 
                                         6.3952 ns    3.40577 ns    10.5777 ns 
Random::Normal                                 100           318      2.544 ms 
                                        89.6539 ns    88.9141 ns    90.5436 ns 
                                        4.10861 ns    3.33656 ns    5.50594 ns 
```
I then gave it the hoisting treatment that we do for `Real()` and `Int()`, where we construct the distribution once and then just sample from it:
```
benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Random::Normal                                 100           597     2.5074 ms 
                                        41.8409 ns    41.5003 ns    42.2892 ns 
                                        1.98118 ns    1.57448 ns    3.10813 ns 
Random::Normal                                 100           590      2.537 ms 
                                        41.9525 ns    41.6092 ns    42.3754 ns 
                                        1.93061 ns    1.61017 ns    2.80822 ns 
Random::Normal                                 100           603     2.5326 ms 
                                        42.0042 ns    41.6727 ns    42.3959 ns 
                                        1.83115 ns    1.54902 ns    2.49926 ns 
Random::Normal                                 100           616     2.5872 ms 
                                        44.0293 ns    43.1027 ns     45.333 ns 
                                        5.54084 ns    4.26149 ns    7.24977 ns 
```

If I add your method and invoke it for the same distribution (0, 1):
```
benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Random::BoxMuller(0, 1)                        100           489      2.445 ms 
                                        50.9498 ns    50.5648 ns    51.5796 ns 
                                        2.46031 ns    1.70163 ns    4.46708 ns 
Random::BoxMuller(0, 1)                        100           513      2.565 ms 
                                        50.8402 ns     50.442 ns    51.5338 ns 
                                        2.62579 ns    1.59575 ns    4.51392 ns 
Random::BoxMuller(0, 1)                        100           475      2.375 ms 
                                        54.6851 ns    52.5031 ns    58.4011 ns 
                                        14.1318 ns    9.71617 ns    21.5508 ns 
Random::BoxMuller(0, 1)                        100           493      2.465 ms 
                                         52.062 ns    51.0577 ns    54.0695 ns 
                                        6.95554 ns    3.87904 ns    11.2491 ns 
```

To summarize: your custom Box Muller implementation is _more complex_ than using the standard library and also slower. We should continue to use the standard library instead of reimplementing a foundational wheel.

If you have discovered that we truly need the performance gains to avoid frame drops and stutters in _normal gameplay_ (i.e. not with projectiles & submunition counts designed to push the system to its limits), then consider implementing a method that will return a constructed & ready-to-sample `std::normal_distribution<double>` for the desired shape.

Reply:
Would you approve the hoisting behavior of creating the distributions once for master?  If yes, I'll modify my own tests to corroborate this and if they show the same I would have no arguments against the std:: implementation.

Reply:
> If you have discovered that we truly need the performance gains to avoid frame drops and stutters in _normal gameplay_ (i.e. not with projectiles & submunition counts designed to push the system to its limits), then consider implementing a method that will return a constructed & ready-to-sample `std::normal_distribution<double>` for the desired shape.

I think this is referring to using a true normal distribution at all as you mentioned before?

It certainly would be nice if there was an implementation that used an inverted cumulative distribution and was even faster than what I have suggested, but what I have suggested is both better and faster than the previous implementation so it would seem a bit unreasonable to require that I produce an ICD implementation anyway.  Further, if you do approve for the std:: version of Normal sampling to be optimized, the overall method that I've suggested will become even faster.

Reply:
> That being said, would you approve the hoisting behavior of creating the distributions once for master? If yes, I'll modify my own tests to corroborate this and if they show the same I would have no arguments against the std:: implementation.

After correcting the mutex guarding in the Box-Muller method and with the hoisting for the std:: method, my tests yield these results over 100 million iterations:

- Random::Normal vs Random::BoxMuller – 1.0166x faster (Random::Normal is faster)
- Generating usable inaccuracy coefficients, std:: vs Box-Muller – 1.0170x faster (std:: is faster)
- Box-Muller vs Triangular – 1.0687x faster (Box-Muller is faster)
- std:: vs Triangular – 1.0869x faster (std:: is faster)

Overall, these results agree with yours, but I'm not sure why they show smaller differences.  I'm pretty certain I did the hoisting correctly as it's pretty simple (just create the distribution in the `ifndef __linux__` block with mean = 0, sigma = 1).  I'll have to check this through the unit tests side as well.

Reply:
These are my results, still smaller differences than yours but still in agreement.
```
benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Random::StdNormal()                            500           550     41.525 ms 
                                        155.068 ns     153.78 ns    156.633 ns 
                                        16.1646 ns      14.07 ns    18.9391 ns 
                                                                               
Random::StdNormal()                            500           544     41.344 ms 
                                        157.794 ns    156.407 ns    159.358 ns 
                                        16.6817 ns    15.1649 ns    18.3862 ns 
                                                                               
Random::StdNormal(0, 1)                        500           548     41.374 ms 
                                        150.031 ns     149.51 ns    150.678 ns 
                                        6.63819 ns    5.67465 ns    8.22674 ns 
                                                                               
Random::StdNormal(0, 1)                        500           535    41.4625 ms 
                                        153.744 ns    152.871 ns    154.927 ns 
                                        11.5035 ns    9.10915 ns    15.0559 ns 
```
```
benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Random::BoxMullerSample()                      500           557    41.4965 ms 
                                        150.707 ns    150.094 ns    151.533 ns 
                                        8.04336 ns    6.52964 ns    11.2043 ns 
                                                                               
Random::BoxMullerSample()                      500           545    41.6925 ms 
                                        151.102 ns    150.487 ns    151.857 ns 
                                        7.72116 ns    6.62352 ns    9.37263 ns 
                                                                               
Random::BoxMullerSample(0, 1)                  500           556       41.7 ms 
                                        157.715 ns    155.919 ns    159.821 ns 
                                        22.1394 ns    19.6665 ns    25.6335 ns 
                                                                               
Random::BoxMullerSample(0, 1)                  500           536      41.54 ms 
                                        150.838 ns    150.094 ns    151.885 ns 
                                        10.0078 ns    7.91166 ns    13.2615 ns 
```
It could be a difference in our machines/OS.  You can definitely see that overall my computer is much slower than yours.

Reply:
I switched back to the STL implementation because mine is no longer better.  Please note, however, that it's probably only slower due to whatever overhead comes with writing your own function as compared to it alreadt being in the STL; the STL version implements a cache just like I did. ^_^